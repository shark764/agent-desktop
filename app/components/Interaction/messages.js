/*
 * Copyright Â© 2015-2017 Serenova, LLC. All rights reserved.
 */

/*
 * Interaction Messages
 *
 * This contains all the text for the Interaction component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  accept: {
    id: 'app.containers.InteractionsBar.accept',
    defaultMessage: 'Click to accept interaction',
  },
  PSTN: {
    id: 'app.containers.InteractionsBar.PSTN',
    defaultMessage: 'Answer phone to accept interaction',
  },
  wrapup: {
    id: 'app.containers.InteractionsBar.wrapup',
    defaultMessage: 'WRAP UP',
  },
});
