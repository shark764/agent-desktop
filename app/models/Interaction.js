import { List } from 'immutable';

export default class Interaction {
  constructor({ interactionId, channelType, autoAnswer, direction, timeout, toolbarFeatures, recording, customerOnHold, status, customer }) {
    if (channelType === 'voice') {
      // recordingUpdate could be undefined for old flows, but should be enabled in that case
      this.agentRecordingEnabled = toolbarFeatures && toolbarFeatures.recordingUpdate !== false;
      // recording and onHold can have been set by an incoming transfer
      this.recording = recording === true;
      this.onHold = customerOnHold === true;
      this.muted = false;
      this.warmTransfers = new List();
    } else if (channelType === 'sms' || channelType === 'messaging') {
      this.customerAvatarIndex = Math.floor(Math.random() * 17);
    } else if (channelType === 'email') {
      this.customer = customer;
    }
    this.channelType = channelType;
    this.direction = direction;
    this.interactionId = interactionId;
    this.timeout = new Date(timeout).valueOf();
    this.autoAnswer = autoAnswer;
    this.status = status || 'work-offer';
    this.query = {};
    this.wrapupDetails = {
      wrapupUpdateAllowed: false,
      wrapupEnabled: false,
    };
    this.contactAction = 'search';
    this.dispositionDetails = {
      forceSelect: false,
      dispositions: [],
      categories: [],
    };
  }
}
