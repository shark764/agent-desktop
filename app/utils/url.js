/*
 * Copyright Â© 2015-2017 Serenova, LLC. All rights reserved.
 */

/*
  * urlHash
  *
  * Returns a object with the parsed current url hash
  *
  */

export function urlHash() {
  const hashObj = {};
  const hashArr = window.location.hash.substr(1).split('&');
  hashArr.forEach((hashItem) => {
    if (hashItem !== '') {
      const splitHash = hashItem.split('=');
      hashObj[splitHash[0]] = splitHash[1];
    }
  });
  return hashObj;
}

/*
  * clearUrlHash
  */

export function clearUrlHash() {
  window.history.pushState('', document.title, window.location.pathname);
}
