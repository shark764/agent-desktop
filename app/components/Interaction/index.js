/**
*
* Interaction
*
*/

import React from 'react';

import { FormattedMessage } from 'react-intl';
import Radium from 'radium';

import Icon from 'components/Icon';
import Timer from 'components/Timer';
import TimerMinutes from 'components/TimerMinutes';
import Progress from 'components/Progress';

import messages from './messages';

export class Interaction extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      startTime: Date.now(),
      ageSeconds: 0,
      msIntervalId: setInterval(
        () => {
          const ageSeconds = Math.round((Date.now() - this.state.startTime) / 1000);
          if (this.props.status === 'wrapup' && (ageSeconds > this.props.wrapupTime)) {
            SDK.interactions.endWrapup({ interactionId: this.props.interactionId });
            clearInterval(this.state.msIntervalId);
          }
          this.setState({
            ageSeconds,
          });
        }, 1000
      ),
    };
  }

  getRemainingSeconds() {
    switch (this.props.status) {
      case 'pending':
        return Math.max(Math.round((this.props.timeout - this.state.startTime) / 1000) - this.state.ageSeconds, 0);
      default:
        if (this.state.ageSeconds < this.props.targetWrapupTime) {
          return (this.props.targetWrapupTime - this.state.ageSeconds);
        } else if (this.state.ageSeconds < this.props.wrapupTime) {
          return (this.props.wrapupTime - this.state.ageSeconds);
        } else {
          return 0;
        }
    }
  }

  resetAge() {
    this.setState({
      startTime: Date.now(),
      ageSeconds: 0,
    });
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.status !== nextProps.status || this.props.previewText !== nextProps.previewText) {
      this.resetAge();
    }
  }

  componentWillUnmount() {
    clearInterval(this.state.msIntervalId);
  }

  styles = {
    base: {
      cursor: 'pointer',
      padding: '20px 16px 0 16px',
      borderRadius: '3px',
      height: '100px',
      width: '100%',
      borderBottom: '1px solid #141414',
      display: 'flex',
      justifyContent: 'stretch',
    },
    pendingBase: {
      height: '123px',
      borderBottom: 'none',
      marginTop: '11px',
    },
    mainContainer: {
      marginLeft: '8px',
      flexGrow: 1,
      display: 'flex',
      flexDirection: 'column',
    },
    headerContainer: {
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'center',
      flexGrow: 0,
      flexShrink: 0,
    },
    from: {
      fontWeight: '600',
      fontSize: '16px',
      lineHeight: '19px',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      flexGrow: 1,
    },
    previewText: {
      height: '36px',
      lineHeight: '18px',
      marginTop: '5px',
      overflow: 'hidden',
    },
    intentText: {
      color: '#979797',
      fontSize: '12px',
      marginTop: '11px',
    },
    timer: {
      marginLeft: '3px',
      flexGrow: 0,
      flexShrink: 0,
    },
    iconContainer: {
      width: '20px',
      height: '20px',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    wrapupContainer: {
      height: '35px',
      marginTop: '7px',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'space-between',
    },
  };

  getBackgroundColor() {
    if (this.props.status === 'pending') {
      return '#F3F3F3';
    } else if (this.props.selected) {
      return '#0B424E';
    }
    return 'inherit';
  }

  getTimerColor() {
    switch (this.props.status) {
      case 'pending':
        return '#23CEF5';
      case 'wrapup':
        if (this.props.targetWrapupTime && (this.state.ageSeconds >= this.props.targetWrapupTime)) {
          return '#FE4565';
        }
        return '#23CEF5';
      default:
        return 'white';
    }
  }

  getTimer() {
    switch (this.props.status) {
      case 'wrapup':
      case 'pending':
        return this.getRemainingSeconds();
      default:
        switch (this.props.channelType) {
          case 'sms':
          case 'messaging':
          case 'email':
            return <TimerMinutes seconds={this.state.ageSeconds} />;
          case 'voice':
          default:
            return <Timer />;
        }
    }
  }

  render() {
    return (
      <div
        id={`${this.props.status}InteractionContainer-${this.props.interactionId}`}
        className={`${this.props.status}InteractionContainer`}
        style={[this.styles.base, { backgroundColor: this.getBackgroundColor() }, this.props.status === 'pending' ? this.styles.pendingBase : {}]}
        key={this.props.interactionId}
        onClick={this.props.onClick}
        disabled={this.props.selected}
      >
        <div style={this.styles.iconContainer}>
          <Icon name={this.props.icon} />
        </div>
        <div style={this.styles.mainContainer}>
          <div style={this.styles.headerContainer}>
            <div style={this.styles.from}>
              {this.props.from}
            </div>
            <div style={[this.styles.timer, { color: this.getTimerColor() }]}>
              {this.getTimer()}
            </div>
          </div>
          {this.props.status === 'wrapup'
            ? <div style={[this.styles.wrapupContainer, { color: this.getTimerColor() }]}>
              <div>
                <FormattedMessage {...messages.wrapup} />
              </div>
              <div>
                <Progress
                  current={this.getRemainingSeconds()}
                  start={this.state.ageSeconds < this.props.targetWrapupTime ? this.props.targetWrapupTime : this.props.wrapupTime - this.props.targetWrapupTime}
                  barColor={this.getTimerColor()}
                  style={{ width: '100%' }}
                />
              </div>
            </div>
            : <div
              className="previewText"
              style={this.styles.previewText}
              dangerouslySetInnerHTML={{ __html: this.props.previewText }}
            ></div>}
          {this.props.status === 'pending'
            ? <div style={this.styles.intentText}>
              <FormattedMessage {...messages.accept} />
            </div>
            : undefined
          }
        </div>
      </div>
    );
  }
}

Interaction.propTypes = {
  interactionId: React.PropTypes.string,
  from: React.PropTypes.string,
  previewText: React.PropTypes.string,
  channelType: React.PropTypes.string,
  icon: React.PropTypes.string,
  timeout: React.PropTypes.number,
  targetWrapupTime: React.PropTypes.number,
  wrapupTime: React.PropTypes.number,
  status: React.PropTypes.oneOf(['pending', 'active', 'wrapup']).isRequired,
  selected: React.PropTypes.bool,
  onClick: React.PropTypes.func,
};

export default Radium(Interaction);
