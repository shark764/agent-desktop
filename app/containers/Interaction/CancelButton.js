/*
 * Copyright Â© 2015-2017 Serenova, LLC. All rights reserved.
 */

/*
 *
 * CancelButton
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Radium from 'radium';

import Button from 'components/Button';

import { cancelClickToDial } from 'containers/AgentDesktop/actions';

import messages from './messages';

const styles = {
  cancelInteractionBtn: {
    margin: '10px 0 0 auto',
    padding: '.25em .5em',
    maxWidth: '65px',
    pointerEvents: 'auto',
  },
};

export class CancelButton extends React.Component {
  cancelInteraction = (e) => {
    // adding this to prevent other events from bubbling up - namely the
    // event to start the interaction which sits on the same div as the button
    e.stopPropagation();
    this.props.cancelClickToDial(this.props.interaction.interactionId);
  };

  render() {
    if (
      this.props.interaction.direction === 'outbound' &&
      this.props.interaction.channelType === 'voice' &&
      this.props.interaction.status === 'work-initiated' &&
      this.props.interaction.initiatedByCurrentAgent
    ) {
      return (
        <Button
          id="cancelInteractionBeforeActive"
          type="primaryRed"
          text={messages.cancelInteraction}
          onClick={this.cancelInteraction}
          style={[styles.cancelInteractionBtn, this.props.style]}
        />
      );
    } else {
      return null;
    }
  }
}

function mapDispatchToProps(dispatch) {
  return {
    cancelClickToDial: (input) => dispatch(cancelClickToDial(input)),
    dispatch,
  };
}

CancelButton.propTypes = {
  interaction: PropTypes.shape({
    interactionId: PropTypes.string,
    direction: PropTypes.string,
    channelType: PropTypes.string,
    status: PropTypes.string,
    initiatedByCurrentAgent: PropTypes.bool,
  }).isRequired,
  style: PropTypes.object,
  cancelClickToDial: PropTypes.func.isRequired,
};

export default connect(null, mapDispatchToProps)(Radium(CancelButton));
