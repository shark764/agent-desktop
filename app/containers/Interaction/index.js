/*
 * Copyright Â© 2015-2017 Serenova, LLC. All rights reserved.
 */

/**
*
* Interaction
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import Radium from 'radium';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import Dotdotdot from 'react-dotdotdot';

import ErrorBoundary from 'components/ErrorBoundary';

import Icon from 'components/Icon';
import Timer from 'components/Timer';
import TimerMinutes from 'components/TimerMinutes';
import Progress from 'components/Progress';
import Button from 'components/Button';
import { cancelClickToDial } from 'containers/AgentDesktop/actions';

import { selectAwaitingDisposition } from 'containers/AgentDesktop/selectors';
import { selectActiveExtension } from 'containers/AgentStatusMenu/selectors';

import messages from './messages';

const styles = {
  base: {
    cursor: 'pointer',
    padding: '20px 16px 0 16px',
    borderRadius: '3px',
    height: '108px',
    width: '100%',
    borderBottom: '1px solid #141414',
    display: 'flex',
    justifyContent: 'stretch',
    backgroundColor: 'inherit',
  },
  pstnBase: {
    cursor: 'default',
  },
  pendingBase: {
    borderBottom: 'none',
    marginTop: '11px',
    backgroundColor: '#F3F3F3',
  },
  selectedBase: {
    backgroundColor: '#0B424E',
  },
  newInteractionBase: {
    height: '60px',
  },
  mainContainer: {
    marginLeft: '8px',
    flexGrow: 1,
    display: 'flex',
    flexDirection: 'column',
    maxWidth: '230px',
    pending: {
      maxWidth: '210px',
    },
  },
  headerContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    flexGrow: 0,
    flexShrink: 0,
  },
  from: {
    fontWeight: '600',
    fontSize: '16px',
    lineHeight: '19px',
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    flexGrow: 1,
  },
  previewText: {
    height: '36px',
    lineHeight: '18px',
    marginTop: '5px',
  },
  intentText: {
    color: '#979797',
    fontSize: '12px',
    marginTop: '11px',
    display: 'flex',
    flexDirection: 'column',
  },
  timer: {
    marginLeft: '3px',
    flexGrow: 0,
    flexShrink: 0,
  },
  iconContainer: {
    width: '20px',
    height: '20px',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  wrapupContainer: {
    height: '35px',
    marginTop: '7px',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
  cancelInteractionBtn: {
    margin: '10px 0 0 auto',
    padding: '.25em .5em',
    maxWidth: '65px',
  },
  cancelInteractionInProgress: {
    opacity: '.75',
    pointerEvents: 'none',
    cursor: 'default',
  },
};

export class Interaction extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      startTime: Date.now(),
      ageSeconds: 0,
      msIntervalId: setInterval(() => {
        const ageSeconds = Math.round(
          (Date.now() - this.state.startTime) / 1000
        );
        if (
          this.props.status === 'wrapup' &&
          ageSeconds > this.props.wrapupTime
        ) {
          if (!this.props.awaitingDisposition) {
            CxEngage.interactions.endWrapup({
              interactionId: this.props.interactionId,
            });
            clearInterval(this.state.msIntervalId);
          }
        }
        this.setState({
          ageSeconds,
        });
      }, 1000),
    };
  }

  componentWillReceiveProps(nextProps) {
    if (
      this.props.status !== nextProps.status ||
      this.props.previewText !== nextProps.previewText
    ) {
      this.setState({
        startTime: Date.now(),
        ageSeconds: 0,
      });
    }
  }

  componentWillUnmount() {
    clearInterval(this.state.msIntervalId);
  }

  getRemainingSeconds = () => {
    switch (this.props.status) {
      case 'pending':
        return Math.max(
          Math.round((this.props.timeout - this.state.startTime) / 1000) -
            this.state.ageSeconds,
          0
        );
      default:
        if (this.state.ageSeconds < this.props.targetWrapupTime) {
          return this.props.targetWrapupTime - this.state.ageSeconds;
        } else if (this.state.ageSeconds < this.props.wrapupTime) {
          return this.props.wrapupTime - this.state.ageSeconds;
        } else {
          return 0;
        }
    }
  };

  getTimer = () => {
    switch (this.props.status) {
      case 'wrapup':
      case 'pending':
        return this.getRemainingSeconds();
      default:
        switch (this.props.channelType) {
          case 'sms':
          case 'messaging':
          case 'email':
            return <TimerMinutes seconds={this.state.ageSeconds} />;
          case 'voice':
          default:
            return <Timer />;
        }
    }
  };

  getTimerColor = () => {
    switch (this.props.status) {
      case 'pending':
        return '#23CEF5';
      case 'wrapup':
        if (
          this.props.targetWrapupTime &&
          this.state.ageSeconds >= this.props.targetWrapupTime
        ) {
          return '#FE4565';
        }
        return '#23CEF5';
      default:
        return 'white';
    }
  };

  getPreviewText = () => {
    if (this.props.status === 'wrapup') {
      return (
        <div style={[styles.wrapupContainer, { color: this.getTimerColor() }]}>
          <div>
            <FormattedMessage {...messages.wrapup} />
          </div>
          <div>
            <Progress
              start={
                this.state.ageSeconds < this.props.targetWrapupTime
                  ? this.state.startTime
                  : this.state.startTime + this.props.targetWrapupTime * 1000
              }
              finish={
                this.state.ageSeconds < this.props.targetWrapupTime
                  ? this.state.startTime + this.props.targetWrapupTime * 1000
                  : this.state.startTime + this.props.wrapupTime * 1000
              }
              barColor={this.getTimerColor()}
              style={{ width: '100%' }}
            />
          </div>
        </div>
      );
    } else if (this.props.status === 'work-ended-pending-script') {
      return (
        <div style={[styles.previewText, { fontStyle: 'italic' }]}>
          <FormattedMessage {...messages.pendingScript} />
        </div>
      );
    } else {
      return (
        <Dotdotdot
          clamp={2}
          className="previewText"
          style={styles.previewText}
          title={this.props.previewText}
        >
          <p style={{ margin: 0 }} title={this.props.previewText}>
            {this.props.previewText}
          </p>
        </Dotdotdot>
      );
    }
  };

  cancelInteraction = (e) => {
    // adding this to prevent other events from bubbling up - namely the
    // event to start the interaction which sits on the same div as the button
    e.stopPropagation();
    this.props.cancelClickToDial(this.props.interactionId);
  };

  render() {
    if (this.props.status !== 'creating-new-interaction') {
      const pendingPSTN =
        this.props.activeExtension.type === 'pstn' &&
        this.props.status === 'pending' &&
        this.props.channelType === 'voice';
      const acceptMessage = pendingPSTN ? messages.PSTN : messages.accept;
      return (
        <div
          id={`${this.props.status}InteractionContainer-${this.props
            .interactionId}`}
          className={`${this.props.status}InteractionContainer`}
          style={[
            styles.base,
            this.props.selected && styles.selectedBase,
            this.props.status === 'pending' && styles.pendingBase,
            pendingPSTN && styles.pstnBase,
            this.props.isCanceled && styles.cancelInteractionInProgress,
          ]}
          key={this.props.interactionId}
          onClick={this.props.onClick}
          disabled={this.props.selected}
        >
          <div style={styles.iconContainer}>
            <Icon name={this.props.icon} />
          </div>
          <div
            style={[
              styles.mainContainer,
              styles.mainContainer[this.props.status],
            ]}
          >
            <div style={styles.headerContainer}>
              <div style={styles.from}>
                {this.props.from}
              </div>
              <div style={[styles.timer, { color: this.getTimerColor() }]}>
                {this.getTimer()}
              </div>
            </div>
            {this.getPreviewText()}
            {this.props.status === 'pending'
              ? <div style={styles.intentText}>
                <FormattedMessage {...acceptMessage} />
                {this.props.interactionDirection === 'outbound' &&
                    this.props.channelType === 'voice'
                    ? <Button
                      id="cancelInteractionBeforeActive"
                      type="primaryRed"
                      text={messages.cancelInteraction}
                      style={styles.cancelInteractionBtn}
                      onClick={this.cancelInteraction}
                    />
                    : undefined}
              </div>
              : undefined}
          </div>
        </div>
      );
    } else {
      return (
        <div
          id={`${this.props.status}InteractionContainer-${this.props
            .interactionId}`}
          className={`${this.props.status}InteractionContainer`}
          style={[
            styles.base,
            this.props.selected && styles.selectedBase,
            styles.newInteractionBase,
          ]}
          key={this.props.interactionId}
          onClick={this.props.onClick}
          disabled={this.props.selected}
        >
          <div style={styles.iconContainer} />
          <div
            style={[
              styles.mainContainer,
              styles.mainContainer[this.props.status],
            ]}
          >
            <div style={styles.headerContainer}>
              <div style={styles.from}>
                New Interaction
              </div>
            </div>
          </div>
        </div>
      );
    }
  }
}

Interaction.propTypes = {
  interactionId: PropTypes.string,
  from: PropTypes.string,
  previewText: PropTypes.string,
  channelType: PropTypes.string,
  icon: PropTypes.string,
  timeout: PropTypes.number,
  targetWrapupTime: PropTypes.number,
  wrapupTime: PropTypes.number,
  status: PropTypes.oneOf([
    'pending',
    'active',
    'wrapup',
    'creating-new-interaction',
    'work-ended-pending-script',
  ]).isRequired,
  awaitingDisposition: PropTypes.bool.isRequired,
  selected: PropTypes.bool,
  onClick: PropTypes.func,
  activeExtension: PropTypes.object.isRequired,
  isCanceled: PropTypes.bool,
  interactionDirection: PropTypes.string,
  cancelClickToDial: PropTypes.func,
};

const mapStateToProps = (state, props) => ({
  awaitingDisposition: selectAwaitingDisposition(state, props),
  activeExtension: selectActiveExtension(state, props),
});

function mapDispatchToProps(dispatch) {
  return {
    cancelClickToDial: (interactionId) =>
      dispatch(cancelClickToDial(interactionId)),
    dispatch,
  };
}

export default ErrorBoundary(
  connect(mapStateToProps, mapDispatchToProps)(Radium(Interaction))
);
