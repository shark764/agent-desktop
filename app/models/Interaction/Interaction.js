/*
 * Copyright Â© 2015-2017 Serenova, LLC. All rights reserved.
 */

import { List, Map, fromJS } from 'immutable';
import { timeSince } from 'utils/time';
import { isToolbar } from 'utils/url';

export default class Interaction {
  constructor({
    interactionId,
    channelType,
    autoAnswer,
    direction,
    timeout,
    toolbarFeatures,
    status,
    customer,
    contact,
    hideNewInteractionPanelOnWorkAccepted,
    initiatedByCurrentAgent,
    contactMode,
    isSidePanelCollapsed,
    subject,
    timeAccepted,
    callbackRequest,
    popUri,
  }) {
    if (channelType === 'voice') {
      // recordingUpdate could be undefined for old flows, but should be enabled in that case
      this.agentRecordingEnabled =
        toolbarFeatures && toolbarFeatures.recordingUpdate !== false;
      this.muted = false;
      this.togglingRecording = false;
      if (direction === 'agent-initiated' || direction === 'outbound') {
        this.customerConnected = false;
      } else {
        this.customerConnected = true;
      }
    } else if (channelType === 'sms' || channelType === 'messaging') {
      this.customerAvatarIndex = Math.floor(Math.random() * 17);
      this.messageHistory = new List();
      this.currentMessage = '';
      this.isCopied = false;
    }
    if (channelType === 'email' || channelType === 'sms') {
      this.customer = customer;
    }
    if (channelType === 'email' && direction === 'agent-initiated') {
      this.emailReply = new Map({
        message: '',
        tos: new List([
          {
            address: customer,
            name: customer,
          },
        ]),
        ccs: new List(),
        bccs: new List(),
        subject: '',
        attachments: new List(),
      });
    }
    this.isEnding = false;
    this.isMuting = false;
    this.isHolding = false;
    this.warmTransfers = new List();
    this.customFieldsCollapsed = true;
    this.channelType = channelType;
    this.direction = direction;
    this.interactionId =
      interactionId || `${direction}-${channelType}-${customer}`;
    this.contactMode = contactMode || 'search';
    this.timeout = new Date(timeout).valueOf();
    this.autoAnswer = autoAnswer;
    this.status = status || 'work-offer';
    this.wrapupDetails = new Map({
      wrapupUpdateAllowed: false,
      wrapupEnabled: false,
    });
    this.dispositionDetails = new Map({
      forceSelect: false,
      dispositions: new List(),
      categories: new List(),
      selected: new List(),
    });
    this.note = new Map({
      body: '',
      title: '',
      notesPanelHeight: isToolbar() ? 180 : 300,
    });
    if (contact) {
      this.contact = fromJS(contact);
    }
    this.query = new Map();
    this.isSidePanelCollapsed =
      isSidePanelCollapsed !== undefined ? isSidePanelCollapsed : true;
    this.selectedSidePanelTab = 'info';
    if (hideNewInteractionPanelOnWorkAccepted !== undefined) {
      this.hideNewInteractionPanelOnWorkAccepted = hideNewInteractionPanelOnWorkAccepted;
    }
    if (initiatedByCurrentAgent !== undefined) {
      this.initiatedByCurrentAgent = initiatedByCurrentAgent;
    }
    this.activeContactForm = fromJS(activeContactFormBlank);
    if (subject) {
      this.subject = subject;
    }
    if (timeAccepted) {
      this.timeAccepted = timeAccepted;
    }

    let notifications = new List();
    if (callbackRequest) {
      notifications = notifications.push(
        new Map({
          messageKey: 'callbackRequest',
          messageValues: new Map({
            callbackNumber: callbackRequest.callbackNumber,
            waitingFor: timeSince(callbackRequest.callbackRequestedTime),
          }),
          isDimissable: true,
        })
      );
    }
    this.notifications = notifications;
    if (popUri) {
      this.popUri = popUri;
    }
  }
}

export const activeContactFormBlank = {
  formIsDirty: false,
  formIsValid: false,
  contactForm: {},
  formErrors: {},
  showErrors: {},
  unusedFields: {},
  selectedIndexes: {},
  editingContacts: [],
  saveLoading: false,
};
