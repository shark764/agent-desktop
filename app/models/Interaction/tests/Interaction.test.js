import { List } from 'immutable';
import { isToolbar } from 'utils/url';
import Interaction from '../Interaction';

Date.now = jest.fn(() => 0);
Math.random = jest.fn(() => 0.99);

jest.mock('utils/url');
isToolbar.mockImplementation(() => false);

describe('Interaction', () => {
  describe('with all the params', () => {
    const interaction = new Interaction({
      interactionId: 'mock-interaction-id',
      channelType: 'mock-channel-type',
      autoAnswer: true,
      direction: 'mock-direction',
      timeout: 0,
      status: 'mock-status',
      contact: {
        id: 'mock-contact-id',
      },
      hideNewInteractionPanelOnWorkAccepted: true,
      initiatedByCurrentAgent: true,
      contactMode: 'mock-contact-mode',
      isSidePanelCollapsed: false,
      subject: 'mock-subject',
      timeAccepted: 'mock-time-accepted',
      callbackRequest: {
        callbackNumber: 'mock-callback-number',
        callbackRequestedTime: 0,
      },
    });
    it('adds the params to the object', () => {
      expect(interaction).toMatchSnapshot();
    });
  });

  describe('with no params', () => {
    const interaction = new Interaction({});
    it('sets defaults, ignores appropriate undefineds', () => {
      expect(interaction).toMatchSnapshot();
    });
  });

  describe('in toolbar mode', () => {
    beforeEach(() => {
      isToolbar.mockImplementationOnce(() => true);
    });
    it('sets notesPanelHeight to the correct value', () => {
      const interaction = new Interaction({});
      expect(interaction.note.get('notesPanelHeight')).toMatchSnapshot();
    });
  });

  describe('voice', () => {
    const interaction = new Interaction({
      channelType: 'voice',
      toolbarFeatures: {
        recordingUpdate: true,
      },
    });
    it('sets muted to false', () => {
      expect(interaction.muted).toEqual(false);
    });
    it('sets agentRecordingEnabled to value passed in', () => {
      expect(interaction.agentRecordingEnabled).toEqual(true);
    });
  });

  describe('sms', () => {
    const interaction = new Interaction({
      channelType: 'sms',
      customer: 'mock-customer',
    });
    it('sets customerAvatarIndex to a random value between 0 and 16', () => {
      expect(interaction.customerAvatarIndex).toEqual(16);
    });
    it('sets messageHistory to an empty list', () => {
      expect(interaction.messageHistory).toEqual(new List());
    });
    it('sets customer to the value passed in', () => {
      expect(interaction.customer).toEqual('mock-customer');
    });
    it('sets currentMessage to ""', () => {
      expect(interaction.currentMessage).toEqual('');
    });
  });

  describe('messaging', () => {
    const interaction = new Interaction({
      channelType: 'messaging',
    });
    it('sets customerAvatarIndex to a random value between 0 and 16', () => {
      expect(interaction.customerAvatarIndex).toEqual(16);
    });
    it('sets messageHistory to an empty list', () => {
      expect(interaction.messageHistory).toEqual(new List());
    });
    it('sets currentMessage to ""', () => {
      expect(interaction.currentMessage).toEqual('');
    });
  });

  describe('email', () => {
    const interaction = new Interaction({
      channelType: 'email',
      customer: 'mock-customer',
    });
    it('sets customer to the value passed in', () => {
      expect(interaction.customer).toEqual('mock-customer');
    });
  });

  describe('outbound email', () => {
    const interaction = new Interaction({
      channelType: 'email',
      direction: 'agent-initiated',
    });
    it('sets emailReply', () => {
      expect(interaction.emailReply).toMatchSnapshot();
    });
  });

  describe('outbound voice', () => {
    const interaction = new Interaction({
      channelType: 'voice',
      direction: 'agent-initiated',
    });
    it('sets customerConnected', () => {
      expect(interaction.customerConnected).toBe(false);
    });
  });
});
