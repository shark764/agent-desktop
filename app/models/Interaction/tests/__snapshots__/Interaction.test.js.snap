// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interaction in toolbar mode sets notesPanelHeight to the correct value 1`] = `180`;

exports[`Interaction outbound ani sets a custom field on our custom field list 1`] = `
Immutable.List [
  Immutable.Map {
    "id": "outboundAniLabel",
    "isLocalized": true,
    "value": "label 1 (value1)",
  },
]
`;

exports[`Interaction outbound ani sets a empty custom field list 1`] = `Immutable.List []`;

exports[`Interaction outbound email sets emailReply 1`] = `
Immutable.Map {
  "message": "",
  "bccInput": "",
  "ccInput": "",
  "toInput": "",
  "attachments": Immutable.List [],
  "bccs": Immutable.List [],
  "ccs": Immutable.List [],
  "tos": Immutable.List [
    Immutable.Map {
      "address": undefined,
      "name": undefined,
    },
  ],
  "subjectInput": "",
}
`;

exports[`Interaction with all the params adds the params to the object 1`] = `
Interaction {
  "activeContactForm": Immutable.Map {
    "contactForm": Immutable.Map {},
    "formIsDirty": false,
    "selectedIndexes": Immutable.Map {},
    "showErrors": Immutable.Map {},
    "editingContacts": Immutable.List [],
    "formIsValid": false,
    "unusedFields": Immutable.Map {},
    "formErrors": Immutable.Map {},
    "saveLoading": false,
  },
  "autoAnswer": true,
  "channelType": "mock-channel-type",
  "contact": Immutable.Map {
    "id": "mock-contact-id",
  },
  "contactMode": "mock-contact-mode",
  "customFields": Immutable.List [],
  "customFieldsCollapsed": true,
  "direction": "mock-direction",
  "dispositionDetails": Immutable.Map {
    "forceSelect": false,
    "dispositions": Immutable.List [],
    "categories": Immutable.List [],
    "selected": Immutable.List [],
  },
  "hideNewInteractionPanelOnWorkAccepted": true,
  "initiatedByCurrentAgent": true,
  "interactionId": "mock-interaction-id",
  "isColdTransferring": false,
  "isEnding": false,
  "isHolding": false,
  "isMuting": false,
  "isSidePanelCollapsed": false,
  "note": Immutable.Map {
    "body": "",
    "title": "",
    "notesPanelHeight": 300,
  },
  "notifications": Immutable.List [
    Immutable.Map {
      "messageKey": "callbackRequest",
      "messageValues": Immutable.Map {
        "callbackNumber": "mock-callback-number",
        "waitingFor": "0s",
      },
      "isDimissible": true,
    },
  ],
  "query": Immutable.Map {},
  "selectedSidePanelTab": "info",
  "status": "mock-status",
  "subject": "mock-subject",
  "timeAccepted": "mock-time-accepted",
  "timeout": 0,
  "transferringInConference": false,
  "warmTransfers": Immutable.List [],
  "wrapupDetails": Immutable.Map {
    "wrapupUpdateAllowed": false,
    "wrapupEnabled": false,
  },
}
`;

exports[`Interaction with no params sets defaults, ignores appropriate undefineds 1`] = `
Interaction {
  "activeContactForm": Immutable.Map {
    "contactForm": Immutable.Map {},
    "formIsDirty": false,
    "selectedIndexes": Immutable.Map {},
    "showErrors": Immutable.Map {},
    "editingContacts": Immutable.List [],
    "formIsValid": false,
    "unusedFields": Immutable.Map {},
    "formErrors": Immutable.Map {},
    "saveLoading": false,
  },
  "autoAnswer": undefined,
  "channelType": undefined,
  "contactMode": "search",
  "customFields": Immutable.List [],
  "customFieldsCollapsed": true,
  "direction": undefined,
  "dispositionDetails": Immutable.Map {
    "forceSelect": false,
    "dispositions": Immutable.List [],
    "categories": Immutable.List [],
    "selected": Immutable.List [],
  },
  "interactionId": "undefined-undefined-undefined",
  "isColdTransferring": false,
  "isEnding": false,
  "isHolding": false,
  "isMuting": false,
  "isSidePanelCollapsed": true,
  "note": Immutable.Map {
    "body": "",
    "title": "",
    "notesPanelHeight": 300,
  },
  "notifications": Immutable.List [],
  "query": Immutable.Map {},
  "selectedSidePanelTab": "info",
  "status": "work-offer",
  "timeout": NaN,
  "transferringInConference": false,
  "warmTransfers": Immutable.List [],
  "wrapupDetails": Immutable.Map {
    "wrapupUpdateAllowed": false,
    "wrapupEnabled": false,
  },
}
`;
